include/boost/histogram/python/axis.hpp

```diff
@@ -87,89 +178,76 @@ decltype(auto) unchecked_bin(const A& ax, bh::axis::index_type i) {
+        },
         ax);
 }
 
-template <class A>
-py::array bins_impl(const A& ax, bool flow) {
-    const bh::axis::index_type underflow
-        = flow && (bh::axis::traits::options(ax) & option::underflow);
-    const bh::axis::index_type overflow
-        = flow && (bh::axis::traits::options(ax) & option::overflow);
-
-    py::array_t<double> result(
-        {static_cast<std::size_t>(ax.size() + underflow + overflow), std::size_t(2)});
-
-    for(auto i = -underflow; i < ax.size() + overflow; ++i) {
-        result.mutable_at(static_cast<std::size_t>(i + underflow), 0) = ax.value(i);
-        result.mutable_at(static_cast<std::size_t>(i + underflow), 1) = ax.value(i + 1);
-    }
-
-    return std::move(result);
-}
-
-template <class... Ts>
-py::array bins_impl(const bh::axis::integer<int, Ts...>& ax, bool flow) {
-    const bh::axis::index_type underflow
-        = flow && (bh::axis::traits::options(ax) & option::underflow);
-    const bh::axis::index_type overflow
-        = flow && (bh::axis::traits::options(ax) & option::overflow);
-
-    py::array_t<int> result(static_cast<std::size_t>(ax.size() + underflow + overflow));
-
-    for(auto i = -underflow; i < ax.size() + overflow; ++i)
-        result.mutable_at(static_cast<std::size_t>(i)) = ax.value(i);
-
-    return std::move(result);
-}
-
-template <class... Ts>
-py::array bins_impl(const bh::axis::category<int, Ts...>& ax, bool flow) {
-    static_assert(!(std::decay_t<decltype(ax)>::options() & option::underflow),
-                  "discrete axis never has underflow");
-
-    const bh::axis::index_type overflow
-        = flow && (bh::axis::traits::options(ax) & option::overflow);
-
-    py::array_t<int> result(static_cast<std::size_t>(ax.size() + overflow));
-
-    for(auto i = 0; i < ax.size() + overflow; ++i)
-        result.mutable_at(static_cast<std::size_t>(i)) = ax.value(i);
-
-    return std::move(result);
-}
-
-template <class... Ts>
-py::array bins_impl(const bh::axis::category<std::string, Ts...>& ax, bool flow) {
-    using namespace pybind11::literals;
-
-    static_assert(!(std::decay_t<decltype(ax)>::options() & option::underflow),
-                  "discrete axis never has underflow");
-
-    const bh::axis::index_type overflow
-        = flow && (bh::axis::traits::options(ax) & option::overflow);
-
-    const auto n = max_string_length(ax);
-    // TODO: this should return unicode
-    py::array result(py::dtype("S" + std::to_string(n + 1)), ax.size() + overflow);
-
-    for(auto i = 0; i < ax.size() + overflow; i++) {
-        auto sout     = static_cast<char*>(result.mutable_data(i));
-        const auto& s = ax.value(i);
-        std::copy(s.begin(), s.end(), sout);
-        sout[s.size()] = 0;
-    }
-
-    return result;
-}
-
-/// Utility to convert bins of axis to numpy array
-template <class A>
-py::array bins(const A& ax, bool flow = false) {
-    // this indirection is needed by pybind11
-    return bins_impl(ax, flow);
-}
+// template <class A>
```

Comment:
Why is there lots of commented out code here?

Reply:
I noticed it is not used any more, which surprised me. To indicate this, I uncommented the code. Why are we not using `.bins`?

Reply:
I wrote this code, but I don't remember why it is not used any more. Apparently, we don't need it. It is probably for the best, since it is redundant: if you want all bins you can do `list(axis)` or `np.array(axis)`.

Reply:
Looking at it again. With `.bins(True)` you also get the extra flow bins.

Reply:
Why is `.edges` a property, it has to be a method, so that you can pass `.edges(True)` to get all edges.

Reply:
Lol, I changed this myself in #121, now why did I do that...

Reply:
I think the reasoning was to consistently hide the flow bins in the axis interface (unless explicitly requested with `ax[bh.underflow]` etc.). The axis descriptions and the view into the storage memory have to be synchronized anyway, and the only place where we can offer this kind of synchronisation is in `.to_numpy`, so that would be the one place to request the flow bins. Still makes sense, no?

Reply:
i removed the commented code and added a comment for .edges